### ISSUE TITLE
Security bugs and vulnerabilities in transitive dependencies of spring cloud skipper server

### ISSUE THREAD

**Issue by DanBarL** [spring-cloud/spring-cloud-skipper#908]
Here is the list of issues:

**1. Vulnerability ID: CVE-2019-12402**
Artifact: commons-compress 
Affected version: 1.18 
Fixed in version: 1.19 

**2. Vulnerability ID: CVE-2019-16869**
Artifact: netty-codec-http
Affected version: 4.1.36.Final
Fixed in version:4.1.42.Final

**3. Vulnerability ID: CVE-2019-11269** 
Artifact: spring-security-oauth2
Affected version: 2.3.4.RELEASE
Fixed in version: 2.3.5 

**4. Vulnerability ID: CVE-2019-3778**
Artifact: spring-security-oauth2
Affected version: 2.3.4.RELEASE 
Fixed in version: 2.3.6 

Could you please update the dependencies to the latest/secured versions?

Regards,

Dan

**Comment by shalomyasap** [comment]
Hi @jvalkeal,

Do you have any estimation when those security issues will be upgraded? 

Thanks,
Shalom (SAP)

**Comment by shalomyasap** [comment]
Hi, 

There are additional two security vulnerabilities: 
5. Artifact bcprov-jdk15on :
    Affected version: 1.60
    Fixed in version: 1.64

6. Artifact guava:
     Affected version: 20.0     
     Fixed in version: 24.1.1

**Comment by sabbyanandan** [comment]
Quick update on this.

All the listed dependencies were never directly used by Skipper or SCDF. They are all transitive dependencies coming from Reactor, Boot, CF-JavaClient, Java CF ENV, and other projects.

To address this, however, we have had work with all the dependent teams to have them update the listed dependencies. With that, we now have all the latest upstream dependencies pulled in Skipper/SCDF, as well. We plan to release the upgraded stack next week.

**Comment by dbahatSAP** [comment]
Excellent! Thanks for the update!

**Comment by shalomyasap** [comment]
Thank you for the update!!

**Comment by dbahatSAP** [comment]
@sabbyanandan I can see some of the issues were resolved, but the updated java CF env change was reverted (https://github.com/spring-cloud/spring-cloud-skipper/commit/93d041b1498278d627778169e7db7ba20ee4949e), so spring security bugs are still present in the latest build (CVE-2019-11269, CVE-2019-3778).

Also netty-codec-http2-4.1.36 seems still present in spring-cloud-starter-common-security-config-web 1.2.0.RC1 (which is associated with bugs CVE-2019-9515, CVE-2019-9512, CVE-2019-9514, CVE-2019-9518), as well as bcpkix-jdk15on 1.6 in spring-cloud-starter-config 2.2.0.RC1 (CVE-2019-17359) and Guava 20.0 in spring-cloud-deployer-resource-maven.2.1.0.RC1 (CVE-2018-10237).

Should we create an additional issue to track the remaining transitive dependencies security issues?

**Comment by sabbyanandan** [comment]
Unfortunately, we have had a few backward incompatibility changes from the upstream dependencies to deal with around the 2.3 RC, so we have had to back-off. Once again, these dependencies that you list aren't directly used by SCDF or Skipper, but we have to deal with it.

There are CFEnv changes inflight to make that work with Spring Cloud Services, so we will have to wait for all that to stabilize. Same also with Spring Security related components. We may end up explicitly defining them (albeit not directly used) in our projects so that we don't have to wait and slip for many more weeks. Feel free to subscribe to: https://github.com/spring-cloud/spring-cloud-dataflow/issues/3634, for updates.

Lastly, what you see re: Guave is already resolved in Spring Cloud Deployer. The current GA releases of SCDF and Skipper will line it up correctly.

**Comment by dbahatSAP** [comment]
@sabbyanandan Thanks for the help!
Looking in Skipper 2.3 release, the spring-security and guava issues were fixed, but there are still multiple transitive security issues originating from the use of io.projectreactor.netty inside spring-cloud-common-security-config-web.
Here's the list:

[ERROR] 1: [filename=netty-codec-http-4.1.36.Final.jar, scope=COMPILE, transitive=true, wellknownSha1=true, isAffectedVersionConfirmed=true, bug=CVE-2019-16869]
[ERROR] 2: [filename=netty-codec-http2-4.1.36.Final.jar, scope=COMPILE, transitive=true, wellknownSha1=true, isAffectedVersionConfirmed=true, bug=CVE-2019-9512]
[ERROR] 3: [filename=bcprov-jdk15on-1.60.jar, scope=COMPILE, transitive=true, wellknownSha1=true, isAffectedVersionConfirmed=true, bug=CVE-2019-17359]
[ERROR] 4: [filename=netty-codec-http2-4.1.36.Final.jar, scope=COMPILE, transitive=true, wellknownSha1=true, isAffectedVersionConfirmed=true, bug=CVE-2019-9515]
[ERROR] 5: [filename=netty-codec-http2-4.1.36.Final.jar, scope=COMPILE, transitive=true, wellknownSha1=true, isAffectedVersionConfirmed=true, bug=CVE-2019-9514]
[ERROR] 6: [filename=netty-codec-http2-4.1.36.Final.jar, scope=COMPILE, transitive=true, wellknownSha1=true, isAffectedVersionConfirmed=true, bug=CVE-2019-9518]
[ERROR] 7: [filename=reactor-netty-0.8.8.RELEASE.jar, scope=COMPILE, transitive=true, wellknownSha1=true, isAffectedVersionConfirmed=true, bug=CVE-2019-11284]

Is there an option to upgrade the reactor netty version? (It seems like a direct dependency of the common security config component)

**Comment by jvalkeal** [comment]
@dbahatSAP Not sure what tool you're using to list these.

Skipper `2.2.1` (there's no `2.3`) uses spring-cloud-common-security-config-web `1.2.1` which gets these from boot `2.2.1`:
```
org.springframework.cloud:spring-cloud-common-security-config-web:jar:1.2.1.RELEASE
+- io.projectreactor.netty:reactor-netty:jar:0.9.1.RELEASE:compile
   +- io.netty:netty-codec-http:jar:4.1.43.Final:compile
      +- io.netty:netty-common:jar:4.1.43.Final:compile
      +- io.netty:netty-buffer:jar:4.1.43.Final:compile
      +- io.netty:netty-transport:jar:4.1.43.Final:compile
         \- io.netty:netty-resolver:jar:4.1.43.Final:compile
      \- io.netty:netty-codec:jar:4.1.43.Final:compile
   +- io.netty:netty-codec-http2:jar:4.1.43.Final:compile
   +- io.netty:netty-handler:jar:4.1.43.Final:compile
   +- io.netty:netty-handler-proxy:jar:4.1.43.Final:compile
      \- io.netty:netty-codec-socks:jar:4.1.43.Final:compile
   \- io.netty:netty-transport-native-epoll:jar:linux-x86_64:4.1.43.Final:compile
      \- io.netty:netty-transport-native-unix-common:jar:4.1.43.Final:compile
```

What comes for `bcprov-jdk15on-1.60.jar` which is coming from `org.springframework.security:spring-security-rsa:jar:1.0.7.RELEASE`, doesn't seem to contain cve's as per https://www.bouncycastle.org/releasenotes.html regression happened in `1.63` which were then fixed in `1.64`.

**Comment by jvalkeal** [comment]
Anyway looks like next `spring-cloud` release will have changes pulling in `bcprov-jdk15on-1.64`.
